---
title: "EnsemblGenomes"
author: 
  - name: Jacques Serizay
    affiliation:
    - Institut Pasteur
    email: jacques.serizay@pasteur.fr
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('EnsemblGenomes')`"
vignette: >
  %\VignetteIndexEntry{test}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    width = 180, 
    crop = NULL
)
```

# Introduction 

`EnsemblGenomes` scrapes [ensembl.org] and [ensemblgenomes.org] FTP 
servers to locate genome reference fasta files and genome annotation 
gff3 files provided for species supported by Ensembl. As of July 2024, 
this corresponds to more than 300 vertebrate, 300 metazoa, 200 protists, 
150 plants, 1,000 fungi and 30,000 bacteria species. Rather than supporting
`BiocFileCache`, EnsemblGenomes simply intends to retrieve and list URL 
of fasta and gff3 files across Ensembl releases as a plain data frame. 

# Installation

`EnsemblGenomes` package can be installed from Bioconductor using the following
command: 

```{r eval = FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("EnsemblGenomes")
```

# `list_ensembl_*` functions

Four functions are available: 

- `list_ensembl_files()`
- `list_ensembl_species()`
- `list_ensembl_taxa()`
- `list_ensembl_releases()`

```{r}
library(EnsemblGenomes)

list_ensembl_taxa()

list_ensembl_releases(taxon = 'vertebrate')

list_ensembl_species('vertebrate', release = 'release-100')

list_ensembl_files('amphiprion_percula', release = 'release-100')
```

# Session info

```{r}
sessionInfo()
```
